src/compiler/core.ts

```diff
@@ -2628,4 +2628,8 @@ namespace ts {
     export function and<T>(f: (arg: T) => boolean, g: (arg: T) => boolean) {
         return (arg: T) => f(arg) && g(arg);
     }
+
+    export function never(never: never): never {

```

Comment:
I think this should be in `Debug` and should be an `assert`, like `assertNever`. I also think that retaining the custom error messages which were at these sites prior to this is valuable; so it should have an (optional?) argument for the message. It should likely also defer to `Debug.fail` for throwing the error so it can pass the `never` asserting function ass the `stackCrawlMark` so that there are not extra irrelevant stack frames in the error.